zerver/models.py

```diff
@@ -953,6 +953,16 @@ class RealmFilter(models.Model):
     realm: Realm = models.ForeignKey(Realm, on_delete=CASCADE)
     pattern: str = models.TextField()
     url_format_string: str = models.TextField(validators=[URLValidator(), filter_format_validator])
+    render_format_string: str = models.TextField(
+        default="",
+        blank=True,
+        validators=[
+            RegexValidator(
+                regex=r"^([\.\/:\w#?=&;!@~-]*%\(([\w-]+)\)s)*[/\w#?=&;~@-]*$",
```

Comment:
How did you select this validator string?

Reply:
Since `render_format_string` and `url_format_string` will have the same format: `foo/%(id)s/bar` so I picked the validator string used by `url_format_string`:
https://github.com/zulip/zulip/blob/8cd9fc7a7411b29cf1a007fa49cf1d543f77063c/zerver/models.py#L941

A slight difference is that I've added `!` and `@` as the allowed characters. Since GitLab users may want to use it as `!1234` and for commits as `zulip/zulip@abcdefg`.
And since it is not necessary to use all the detected groups in the linkifier `render_format_string`, I've used the `*` qualifier rather than the `+` qualifier.